### PR TITLE
Split Sweden into subzones.

### PR DESCRIPTION
**Pull request by VIKTORVAV99**
This PR splits Sweden to the subzones SE-SE1, SE-SE2, SE-SE3 and SE-SE4 but should NOT be merge until #1397 is unblocked.

TODO:
- [x] Add subzone capacities. (Looks like the source includes small scale gasoline and diesel backups in gas but it's the only one I could find that splits it per bidding zone. We should try and find a better source in the future) 
- [x] Update exchanges.
- [x] Update parsers.
- [x] Change parsers in `zones.json` and add `subZoneNames` to `SE`.
- [x] Add gCO2eq/kWh for unknown sources. See #3915 for reference.
- [x] Update `world.geojson` with SE-SE1, SE-SE2, SE-SE3, SE-SE4 and remove SE. Help Needed!
- [x] Update translation files with new names for SE subzones.
- [x] Update tests.

I know there is already a PR that aims to do the same thing but it has become outdated and need a lot of edits to get in the right shape to be merged, this PR aims to fix that.

PR:s and commits to this PR are welcome!

Fixes: #1397, #2699

### CODE REVIEW

**Comment by gwpicard** [comment]
Thanks for this PR @VIKTORVAV99 — super great to be able to get this one going finally. We're holding off on this one because we don't yet have the ability to have two regional granularities flowtranced at the same time (ie SE and SE subzones). And we don't want to switch to subzones and get rid of SE until we have this.

For more context, see also similar issues for https://github.com/electricitymap/electricitymap-contrib/issues/1397 (Sweden), https://github.com/electricitymap/electricitymap-contrib/issues/3795 (Germany) and https://github.com/electricitymap/electricitymap-contrib/issues/2777 (France)

Once this is implemented (hopefully it will be integrated on the road map this year), we will be in a position to merge this.

But thanks so much for getting the work started on this!

**Comment by VIKTORVAV99** [comment]
@gwpicard I am aware of the current constraints but my goal with this PR is not to get it merged right away but to have everything ready by the point you can flowtrace multiple granularities. Hopefully, I (we?) can maintain this PR until that point and then have a simple merge without much delay.

**Comment by gwpicard** [comment]
> @gwpicard I am aware of the current constraints but my goal with this PR is not to get it merged right away but to have everything ready by the point you can flowtrace multiple granularities. Hopefully, I (we?) can maintain this PR until that point and then have a simple merge without much delay.

Yep, agreed. I don't think we should close it but just wanted to flag it so we're aligned. 

However, we should probably close https://github.com/electricitymap/electricitymap-contrib/issues/1397

**Comment by VIKTORVAV99** [comment]
@gwpicard, Issue #1397 has become outdated and should be closed but it might make sense to open a new issue, partially to prevent someone else from creating a duplicate and to have a place to leave comments not related to the code/changes in this PR.

**Comment by gwpicard** [comment]
> @gwpicard, Issue #1397 has become outdated and should be closed but it might make sense to open a new issue, partially to prevent someone else from creating a duplicate and to have a place to leave comments not related to the code/changes in this PR.

I think we should keep #1397 because it has useful context and past conversations relevant to this PR. But yes, now that it's tagged again here, people can add comments over on the issue rather than on this PR.

**Comment by pierresegonne** [comment]
> @pierresegonne While maintaining this PR I noticed that isort is passing locally but not in the CI, any reason for that?

I guess it might be a version issue? What version do you have locally?

**Comment by VIKTORVAV99** [comment]
> > @pierresegonne While maintaining this PR I noticed that isort is passing locally but not in the CI, any reason for that?
> 
> I guess it might be a version issue? What version do you have locally?

My local version is 5.9.3, the same version that is specified in `pyproject.toml`

**Comment by VIKTORVAV99** [comment]
> Hey @VIKTORVAV99 :)
> 
> When pulling your changes locally and running `poetry run check` isort fails for me. So I guess it's because your isort config is outdated or something along these lines. You could remove and reinstall without cache to double check.
> 
> In the meanwhile, these are the imports in `test_parser.py` that should make isort happy:
> 
> ```python
> import datetime
> import logging
> import pprint
> import time
> 
> import arrow
> import click
> 
> from electricitymap.contrib.parsers.lib.parsers import PARSER_KEY_TO_DICT
> from parsers.lib.quality import (
>     ValidationError,
>     validate_consumption,
>     validate_exchange,
>     validate_production,
> )
> ```
> 
> 

Thanks I'll give it a try.

On a slightly off topic note is there a reason we are importing from `electricitymap.contrib.parsers.lib.parsers` instead of just `parsers.lib.parsers`? (like the one just after)

**Comment by VIKTORVAV99** [comment]
> > Hey @VIKTORVAV99 :)
> > When pulling your changes locally and running `poetry run check` isort fails for me. So I guess it's because your isort config is outdated or something along these lines. You could remove and reinstall without cache to double check.
> > In the meanwhile, these are the imports in `test_parser.py` that should make isort happy:
> > ```python
> > import datetime
> > import logging
> > import pprint
> > import time
> > 
> > import arrow
> > import click
> > 
> > from electricitymap.contrib.parsers.lib.parsers import PARSER_KEY_TO_DICT
> > from parsers.lib.quality import (
> >     ValidationError,
> >     validate_consumption,
> >     validate_exchange,
> >     validate_production,
> > )
> > ```
> 
> Thanks I'll give it a try.
> 
> On a slightly off topic note is there a reason we are importing from `electricitymap.contrib.parsers.lib.parsers` instead of just `parsers.lib.parsers`? (like the one just after)

@pierresegonne turns out when using the import `electricitymap.contrib.parsers.lib.parsers` `isort` identifies it as a `THRIDPARTY` module but when using the shorter `parsers.lib.parsers` import `isort` identifies it as a `FIRSTPARTY` module just as `parsers.lib.quality`.

I don't know if this is affected by the operating system (I think it is since windows don't appears to be able to resolve the longer symbolic import) but I think we should switch to the shorter import just to be safe. I'll open a separate PR for that and we can discuss it more there if there is any issues with going down that path.

**Comment by pierresegonne** [comment]
Hey @VIKTORVAV99, the situation is unlocked on our side, so we can start looking into merging this finally :)

**Comment by VIKTORVAV99** [comment]
> Hey @VIKTORVAV99, the situation is unlocked on our side, so we can start looking into merging this finally :)

Awesome! 🎉

It should only be the geographies that's missing, perhaps you could help with those @Kongkille?

(I can't fit the GiS editing software on my hard drive at the moment... 😅)

**Comment by pierresegonne** [comment]
FYI in #3607 there were some work done on the geometries

**Comment by Kongkille** [comment]
@VIKTORVAV99 I'll try to add the geographies to this PR!

**Comment by VIKTORVAV99** [comment]
> FYI in #3607 there were some work done on the geometries
> 

It where but it used the old geometry system so we sadly can't just copy the changes in that pr. 😕

**Comment by PierreMesure** [comment]
Exciting to see this merged and really interesting to read the discussions about the journey! It seems so absurd to me that the Swedish government decided on splitting the grid in 4 zones without opening all the data related to these areas.

Regarding the GIS data, I can try to have a lot if no one's working on it. Just let me know.

**Comment by VIKTORVAV99** [comment]
> Exciting to see this merged and really interesting to read the discussions about the journey! It seems so absurd to me that the Swedish government decided on splitting the grid in 4 zones without opening all the data related to these areas.
> 
> Regarding the GIS data, I can try to have a lot if no one's working on it. Just let me know.

The data has been available for a while on ENTSO-E and has been released as yearly reports. It has just been some blocking factors that are now resolved.

As for the geographies @Kongkille is working on those but you can always double check with him.

Otherwise I don't think he'd mind if you helped out with the geographies for Ceuta and Melilla in #4318. They are very small zones but it would be nice to get them added all the same.

**Comment by Kongkille** [comment]
Geographies seem to be working now :)

image

It was a bit difficult because our validation on checking gaps has a tendency to produce some odd results, so I've included some code in the geo validator to ensure we don't get stuck on small polygon slivers

image

**Comment by VIKTORVAV99** [comment]
> Geographies seem to be working now :)
> 
> image
> 
> It was a bit difficult because our validation on checking gaps has a tendency to produce some odd results, so I've included some code in the geo validator to ensure we don't get stuck on small polygon slivers
> 
> image
> 
> 
> 
> 

Nice!

I think the actual bidding zone borders are a lot weirder (not straight lines) but this should be more than sufficient enough us to deploy it to the map at least. (we can always improve them later if needed, for example I'd love to add the two great lakes at one point)

**Comment by VIKTORVAV99** [comment]
I'll update the Jest tests to match the new zones config but I think this is ready to merge after that.

**Comment by VIKTORVAV99** [comment]
@Kongkille I updated the Jest test but it's still failing, is it because there is no exchange data between DK-DK2 and SE-SE4 in the mock data?

**Comment by Kongkille** [comment]
> @Kongkille I updated the Jest test but it's still failing, is it because there is no exchange data between DK-DK2 and SE-SE4 in the mock data?

Ah yeah I'll just move the Jest test to another zone. We should actually be able to have both SE and SE-SE* but I don't think we need to address this in the dataReducer test.

**Comment by gwpicard** [comment]
@VIKTORVAV99 

I saw that on SvK the official names for the bidding areas are:

- Luleå (SE1)
- Sundsvall (SE2)
- Stockholm (SE3)
- Malmö (SE4)

[Source](https://www.svk.se/en/national-grid/the-control-room/)

Up to you whether you think we should keep the current names you had, thought it might be nice to align with the official names.

**Comment by VIKTORVAV99** [comment]
> @VIKTORVAV99
> 
> I saw that on SvK the official names for the bidding areas are:
> 
> * Luleå (SE1)
> * Sundsvall (SE2)
> * Stockholm (SE3)
> * Malmö (SE4)
> 
> [Source](https://www.svk.se/en/national-grid/the-control-room/)
> 
> Up to you whether you think we should keep the current names you had, thought it might be nice to align with the official names.

It might be the "official" names but I used the names that [E.on](https://www.eon.se/el/elpriser/aktuella) (a major electricity utility company) used (and they translate way better).

**Comment by PierreMesure** [comment]
In Sweden, these zones are simply called SE1, SE2, SE3, SE4. I can ask my friend at SvK but I've never seen anyone using the city names. I also find it misleading as these zones are much bigger and include other big cities than these 4 (Göteborg, for instance).

**Comment by VIKTORVAV99** [comment]
> In Sweden, these zones are simply called SE1, SE2, SE3, SE4. I can ask my friend at SvK but I've never seen anyone using the city names. I also find it misleading as these zones are much bigger and include other big cities than these 4 (Göteborg, for instance).

I think it's just better to go with the more descriptive name that we can translate.

Maybe you can convince your friend at SvK to change the ones SvK use to the ones E.on use. 😉

**Comment by PierreMesure** [comment]
I actually checked some news articles and many also use the more descriptive names that you (and E.on) use.

File: config/zones.json
Comment (pierresegonne) [inline config/zones.json:4218]: My quick comparison of country wide vs zones capacities shows some discrepencies:
* As noted above, gas is much higher for subzones (1535 vs 698)
* oil is reported at 2251MW for the country while subzones sum up to 905.

For the rest it looks OKish

File: web/src/reducers/dataReducer.test.js
Comment (VIKTORVAV99) [reply web/src/reducers/dataReducer.test.js:102]: Cause the state files used for the mockserver don't have any SE-SE* data yet so it will always fail and if I'm not mistaken we need to merge this to get that data...

File: config/zones.json
Comment (VIKTORVAV99) [reply config/zones.json:4218]: I know, I switched to the official report by SVK (the Swedish TSO) and they appears to be reporting gas together with small scale diesel and gasoline generators, I plan to update this as soon as I find a source that splits it or when I have the time to calculate it myself based on known power plants.

Comment (pierresegonne) [reply config/zones.json:4218]: Alright, we can make a follow up adjustment. Considering the small share of oil / gas in the Swedish mix it's not crucial we get this 100% right from the get go

**Comment by pierresegonne** [comment]
Difference between the `SE` production parser data and the sum of the `SE-SE*` production parser data over 3 months. 

Screenshot 2022-08-10 at 10 52 32

There is 0 difference! 

I guess SvK is doing its job correctly :)

**Comment by pierresegonne** [comment]
Hey @VIKTORVAV99 have you run a quick comparison of the exchanges?
60edfb10-3aa3-47b8-aab0-a2e9c836a98a
c0480089-fa93-4b27-b14c-35dfb45aebc9


When comparing previous exchange data for `FI->SE` to the aggregates of `FI->SE-SE1`, `FI->SE-SE3`

and 

for `NO-NO4->SE` and `NO-NO4->SE-SE2`, `NO-NO4->SE-SE1` it seems like there are some quite significant differences. The rest matches very well (few outliers but that's ok)

**Comment by pierresegonne** [comment]
> or NO-NO4->SE and NO-NO4->SE-SE2, NO-NO4->SE-SE1 it seems like there are some quite significant differences.

The following suggests that `NO-NO4->SE-SE1` was not taken into account at all before. What suggests this is that there is absolutely no difference between the old `NO-NO4->SE` and the new `NO-NO4->SE-SE2`


3e6f2c89-de07-4f67-94d0-904fdb7fd5a5

**Comment by pierresegonne** [comment]
> The following suggests that NO-NO4->SE-SE1 was not taken into account at all before. What suggests this is that there is absolutely no difference between the old NO-NO4->SE and the new NO-NO4->SE-SE2

That's correct, there was a mistake. [This commit](https://github.com/electricitymap/electricitymap-contrib/commit/850163a640847e1323e70facba614c4289a7e265) matched the `NO-NO4->SE` to only `NO-NO4->SE-SE2`, probably because at the time that was the only exchange. 

So basically the exchange data has been wrong since 2019.

**Comment by pierresegonne** [comment]
For `FI` the difference must come from a misalignment between the old source, stattnett, and ENTSO-E. On average it's just a few % difference, (50MW on exchanges that go ~2000MW) so I guess it's fine

**Comment by pierresegonne** [comment]
We're ready to merge this, 

We will do it next week, as it will require a bit of internal work on our side to postprocess some of the historical data we have. 

Great job 💪💪💪

**Comment by VIKTORVAV99** [comment]
> For `FI` the difference must come from a misalignment between the old source, stattnett, and ENTSO-E. On average it's just a few % difference, (50MW on exchanges that go ~2000MW) so I guess it's fine

This could possibly include Åland exchange data as thats technically a part of Finland, I don't know if the statnet parser treated them differently.

However since there is no exchange between Åland and Sweden in entso-e I think that's the most likely explanation, I'll check later and if needed write a wrapper around the entso parser and AX parser to deduct the exchange volume between Sweden and Åland.

File: web/geo/update-world.js
Comment (tonypls) [inline web/geo/update-world.js:14]: Hey @VIKTORVAV99,

I'm looking to use your sliver checking function.

One thing I noticed is area is returned in m^2 and line length in km, I think we should increase the sliver ratio to compensate. From my testing a ratio of 10 successfully identifies all of the 6 slivers I have in my data.

Below is an example of two sliver ratios for a thin and square polygon. I'm not sure exactly how line length is calculated by turf but my example assumes perimeter.

2000m x 0.5m  
Area = 1000m^2

Length = 2.001km

Sliver ratio: 1000m^2 / 2 = 500 m^2/km


1000m x 1000m

Area = 1,000,000m^2

Length = 4km

Sliver ratio: 250,000m^2/km

I think a sliver ratio of 10 is a nice conservative approach here but I'd appreciate your feedback.

Comment (VIKTORVAV99) [reply web/geo/update-world.js:14]: Oh man, you'll have to talk to @Kongkille about all this. He did the geographies and related changes and I haven't had time to dig into the details.

**Comment by pierresegonne** [comment]
> However since there is no exchange between Åland and Sweden in entso-e I think that's the most likely explanation, I'll check later and if needed write a wrapper around the entso parser and AX parser to deduct the exchange volume between Sweden and Åland.

I just verified, and the old `FI->SE` exchange also included the `AX->FI` data, so all good!

**Comment by VIKTORVAV99** [comment]
@pierresegonne I re-ran the Cypress test but this is the part that failed:
```
  1) Ranking Panel
       interacts with details:
     CypressError: Timed out retrying after 4050ms: `cy.click()` failed because this element is detached from the DOM.
```

It has happened a few times on a few different PRs now so I think It's a flaky test.
Just so you are aware.

**Review by pierresegonne — APPROVED** [review]

**Comment by pierresegonne** [comment]
> It has happened a few times on a few different PRs now so I think It's a flaky test.
Just so you are aware.

Interesting, @Kongkille FYI

**Comment by Gazer75** [comment]
>> The following suggests that NO-NO4->SE-SE1 was not taken into account at all before. What suggests this is that there is absolutely no difference between the old NO-NO4->SE and the new NO-NO4->SE-SE2

>That's correct, there was a mistake. [This commit](https://github.com/electricitymap/electricitymap-contrib/commit/850163a640847e1323e70facba614c4289a7e265) matched the NO-NO4->SE to only NO-NO4->SE-SE2, probably because at the time that was the only exchange.

So basically the exchange data has been wrong since 2019.

This is what I pointed out in #2699

**Comment by VIKTORVAV99** [comment]
> > > The following suggests that NO-NO4->SE-SE1 was not taken into account at all before. What suggests this is that there is absolutely no difference between the old NO-NO4->SE and the new NO-NO4->SE-SE2
> 
> > That's correct, there was a mistake. [This commit](https://github.com/electricitymap/electricitymap-contrib/commit/850163a640847e1323e70facba614c4289a7e265) matched the NO-NO4->SE to only NO-NO4->SE-SE2, probably because at the time that was the only exchange.
> 
> > So basically the exchange data has been wrong since 2019.
> 
> This is what I pointed out in #2699
> 
> Did anyone find installed capacity per production type for each SE region? Seems only the country total is listed at ENTSO-E.

I did find yearly reports from SVK but they don't break it down in the same granularity as ENTSO-E and they also seem to include small scale gasoline and diesel generators in gas but it's the best we got right now.

**Comment by martian-maniac** [comment]
When is this going live if I may ask? Sweden has general elections on 11th sept and increased visibility in our electricity system can only be a good thing. It's a hot topic.

**Comment by VIKTORVAV99** [comment]
> When is this going live if I may ask? Sweden has general elections on 11th sept and increased visibility in our electricity system can only be a good thing. It's a hot topic.

There was some unforseen additional work that needed to be done on the backend to get everything working correctly. But it shouldn't be too long.

As a Swede my self I can only say I agree that this is a hot topic. It will be especially intresting to see the carbon diffrence between Southern (SE-SE4) and North Central Sweden (SE-SE2).

**Pull request merged**
